src/v/wasm/engine_probe.cc

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "wasm/engine_probe.h"
+
+#include "metrics/metrics.h"
+#include "prometheus/prometheus_sanitize.h"
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/weak_ptr.hh>
+
+namespace wasm {
+
+namespace internal {
+engine_probe_impl::engine_probe_impl(
+  engine_probe_cache* cache, ss::sstring name)
+  : _name(std::move(name))
+  , _cache(cache) {
+    namespace sm = ss::metrics;
+
+    auto name_label = sm::label("function_name");
+    const std::vector<sm::label_instance> labels = {
+      name_label(_name),
+    };
+    _public_metrics.add_group(
+      prometheus_sanitize::metrics_name("wasm_engine"),
+      {
+        sm::make_gauge(
+          "memory_usage",
+          sm::description("Amount of memory usage for a WebAssembly function"),
+          labels,
+          [this] { return _memory_usage; })
+          .aggregate({ss::metrics::shard_label}),
+        sm::make_gauge(
+          "max_memory",
+          [this] { return _max_memory; },
+          sm::description("Max amount of memory for a WebAssembly function"),
+          labels)
+          .aggregate({ss::metrics::shard_label}),
+      });
+}
+engine_probe_impl::~engine_probe_impl() {
+    // Remove from cache when deleted.
+    _cache->remove_probe(_name);
+}
+void engine_probe_impl::report_memory_usage_delta(int64_t delta) {
+    _memory_usage += delta;
+}
+
+void engine_probe_impl::report_max_memory_delta(int64_t delta) {
+    _max_memory += delta;
+}
+} // namespace internal
+
+engine_probe::engine_probe(ss::lw_shared_ptr<internal::engine_probe_impl> impl)
+  : _impl(std::move(impl)) {}
+
+engine_probe::~engine_probe() {
+    if (!_impl) {
+        return;
+    }
+    _impl->report_max_memory_delta(-int64_t(_last_reported_max_memory));
+    _impl->report_memory_usage_delta(-int64_t(_last_reported_memory_usage));
+}
+
+void engine_probe::report_memory_usage(uint32_t usage) {
+    int64_t delta = int64_t(usage) - int64_t(_last_reported_memory_usage);
+    _impl->report_memory_usage_delta(delta);
+    _last_reported_memory_usage = usage;
+}
+
+void engine_probe::report_max_memory(uint32_t max) {
+    int64_t delta = int64_t(max) - int64_t(_last_reported_max_memory);
+    _impl->report_max_memory_delta(delta);
+    _last_reported_max_memory = max;
+}
+
+engine_probe engine_probe_cache::make_probe(const ss::sstring& name) {
+    {
```

Comment:
Bit said that there is no better pattern for this conditional insert but wouldn't just using `operator[]` be the easiest here? 

Can just check for a default constructed nullptr shared_ptr?

Reply:
I miss some of the `gtl` map helpers from my time at Google, but yes default construction works fine - I've made the switch

Reply:
Sorry I meant you can then also replace the emplace call? Something like:

```
    auto& probe = _probe_cache[name];
    if (!probe) {
        probe = ss::make_lw_shared<internal::engine_probe_impl>(
      this, name);
    }
    return engine_probe(probe);
```

or am I missing something?

Reply:
We store these as raw pointers, and the destructor of the probe removes the map entry (a sneaky pattern), because we want this to be a cache, but to cleanup the probes when all the holders to the impl are deleted.

I'm open to better approaches here!

Reply:
Ah sorry missed that, I think you could still avoid the extra emplace call by rearranging the code a bit.

This sounds like the classical weak_ptr pattern but lw_shared probably doesn't support that.

Reply:
It does support `weak_ptr` but I still need to remove the map entries if you have ideas on how to do that.

ref to weak_ptr: https://github.com/redpanda-data/seastar/blob/245e0ccfa6d58d7e0dca2b4034ce1bc43e39bdc5/include/seastar/core/weak_ptr.hh#L97-L113